File: packages/relay/internal/connections/messages.go
package connections

import (
	"encoding/json"
	"relay/internal/common"

	"github.com/pion/webrtc/v4"
)

// MessageBase is the base type for any JSON message
type MessageBase struct {
	Type    string                 `json:"payload_type"`
	Latency *common.LatencyTracker `json:"latency,omitempty"`
}

type MessageRaw struct {
	MessageBase
	Data json.RawMessage `json:"data"`
}

func NewMessageRaw(t string, data json.RawMessage) *MessageRaw {
	return &MessageRaw{
		MessageBase: MessageBase{
			Type: t,
		},
		Data: data,
	}
}

type MessageLog struct {
	MessageBase
	Level   string `json:"level"`
	Message string `json:"message"`
	Time    string `json:"time"`
}

func NewMessageLog(t string, level, message, time string) *MessageLog {
	return &MessageLog{
		MessageBase: MessageBase{
			Type: t,
		},
		Level:   level,
		Message: message,
		Time:    time,
	}
}

type MessageMetrics struct {
	MessageBase
	UsageCPU        float64 `json:"usage_cpu"`
	UsageMemory     float64 `json:"usage_memory"`
	Uptime          uint64  `json:"uptime"`
	PipelineLatency float64 `json:"pipeline_latency"`
}

func NewMessageMetrics(t string, usageCPU, usageMemory float64, uptime uint64, pipelineLatency float64) *MessageMetrics {
	return &MessageMetrics{
		MessageBase: MessageBase{
			Type: t,
		},
		UsageCPU:        usageCPU,
		UsageMemory:     usageMemory,
		Uptime:          uptime,
		PipelineLatency: pipelineLatency,
	}
}

type MessageICE struct {
	MessageBase
	Candidate webrtc.ICECandidateInit `json:"candidate"`
}

func NewMessageICE(t string, candidate webrtc.ICECandidateInit) *MessageICE {
	return &MessageICE{
		MessageBase: MessageBase{
			Type: t,
		},
		Candidate: candidate,
	}
}

type MessageSDP struct {
	MessageBase
	SDP webrtc.SessionDescription `json:"sdp"`
}

func NewMessageSDP(t string, sdp webrtc.SessionDescription) *MessageSDP {
	return &MessageSDP{
		MessageBase: MessageBase{
			Type: t,
		},
		SDP: sdp,
	}
}

File: packages/server/src/messages.rs
use crate::latency::LatencyTracker;
use serde::{Deserialize, Serialize};
use webrtc::ice_transport::ice_candidate::RTCIceCandidateInit;
use webrtc::peer_connection::sdp::session_description::RTCSessionDescription;

#[derive(Serialize, Deserialize, Debug)]
pub struct MessageBase {
    pub payload_type: String,
    pub latency: Option<LatencyTracker>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct MessageRaw {
    #[serde(flatten)]
    pub base: MessageBase,
    pub data: serde_json::Value,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct MessageLog {
    #[serde(flatten)]
    pub base: MessageBase,
    pub level: String,
    pub message: String,
    pub time: String,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct MessageMetrics {
    #[serde(flatten)]
    pub base: MessageBase,
    pub usage_cpu: f64,
    pub usage_memory: f64,
    pub uptime: u64,
    pub pipeline_latency: f64,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct MessageICE {
    #[serde(flatten)]
    pub base: MessageBase,
    pub candidate: RTCIceCandidateInit,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct MessageSDP {
    #[serde(flatten)]
    pub base: MessageBase,
    pub sdp: RTCSessionDescription,
}
